examples/llm/src/models/param_init_fns.py

```diff
@@ -293,9 +293,116 @@ def xavier_normal_param_init_fn_(module, cfg):
     generic_param_init_fn_(module, cfg, xavier_normal_)
 
 
+def deepnorm_init_fn_(module, cfg):
+    # from DeepNet by Wang et al 2022 https://arxiv.org/abs/2203.00555
+    if cfg.get('verbose') and cfg.get('verbose') > 1:
+        warnings.warn(
+            f'If model has bias parameters they are initialized to 0.')
+
+    beta = (8 * cfg.n_layers)**-0.25
+    fvo_init_fn_ = partial(nn.init.xavier_normal_, gain=beta)
+    qk_init_fn_ = partial(nn.init.xavier_normal_, gain=1.)
+
+    if isinstance(module, nn.Linear):
+        # Linear
+        if hasattr(module, '_fused'):
+            fused_init_helper_(module, fvo_init_fn_)
+        else:
+            fvo_init_fn_(module.weight)
+        if module.bias is not None:
+            torch.nn.init.zeros_(module.bias)
+
+    elif isinstance(module, nn.Embedding):
+        # Embedding
+        if cfg.get('emb_init_std') is not None:
+            std = cfg.get('emb_init_std')
+            if std == 0:
+                warnings.warn(f'Embedding layer initialized to 0.')
+            emb_init_fn_ = partial(torch.nn.init.normal_, mean=0.0, std=std)
+            if cfg.get('verbose', 0) > 1:
+                warnings.warn(
+                    f'Embedding layer initialized using normal distribution with mean=0 and {std=}.'
+                )
+        elif cfg.get('emb_init_uniform_lim') is not None:
+            lim = cfg.get('emb_init_uniform_lim')
+            if isinstance(lim, Sequence):
+                if len(lim) > 2:
+                    raise ValueError(
+                        f'Uniform init requires a min and a max limit. User input: {lim}.'
+                    )
+                if lim[0] == lim[1]:
+                    warnings.warn(f'Embedding layer initialized to {lim[0]}.')
+            else:
+                if lim == 0:
+                    warnings.warn(f'Embedding layer initialized to 0.')
+                lim = [-lim, lim]
+            a, b = lim
+            emb_init_fn_ = partial(torch.nn.init.uniform_, a=a, b=b)
+            if cfg.get('verbose', 0) > 1:
+                warnings.warn(
+                    f'Embedding layer initialized using uniform distribution in range {lim}.'
+                )
+        else:
+            emb_init_fn_ = fvo_init_fn_  # deepnet paper does not mention embedding init
```

Comment:
if it does not mention it, there is a good change they are using normal_(mean=0, std=1) since that is the pytorch default init for emb

Reply:
thanks, updated!